### PR TITLE
IGNITE-20645 Make ComputeJob.execute asynchronous

### PR DESCRIPTION
**Pull request by ptupitsyn**
* Change `R ComputeJob.execute(...)` to `CompletableFuture<R> ComputeJob.executeAsync(...)`
* Update cancellation logic to deal with the resulting future

### CODE REVIEW

File: modules/api/src/main/java/org/apache/ignite/compute/ComputeJob.java [hunk 1]
```diff
@@ -17,18 +17,22 @@
...
  */
+@SuppressWarnings("InterfaceMayBeAnnotatedFunctional")
 public interface ComputeJob<R> {
     /**
      * Executes the job on an Ignite node.
      *
      * @param context The execution context.
      * @param args Job arguments.
-     * @return Job result.
+     * @return Job future. Can be null if the job is synchronous and does not return any result.
```
Comment (isapego) [inline modules/api/src/main/java/org/apache/ignite/compute/ComputeJob.java:35]: Performance reasons?

**Review by isapego — APPROVED** [review]

File: modules/api/src/main/java/org/apache/ignite/compute/ComputeJob.java (hunk 1 above)
Comment (ptupitsyn) [reply modules/api/src/main/java/org/apache/ignite/compute/ComputeJob.java:35]: * Very small performance improvement
* A bit easier for the user at the expense of a one-line check in Ignite code

Do you think we should disallow nulls?

File: modules/compute/src/test/java/org/apache/ignite/internal/compute/loader/JobClassLoaderFactoryTest.java [hunk 2]
```diff
@@ -64,14 +65,14 @@ public void unit1() throws Exception {
             // then classes from the first unit are loaded from the first class loader
             Class<?> clazz1 = classLoader1.loadClass(UNIT_JOB_CLASS_NAME);
             ComputeJob<Integer> job1 = (ComputeJob<Integer>) clazz1.getDeclaredConstructor().newInstance();
-            Integer result1 = job1.execute(null);
-            assertEquals(1, result1);
+            CompletableFuture<Integer> result1 = job1.executeAsync(null);
+            assertEquals(1, result1.join());
```
[the remark this answers is not in the document]
Comment (ptupitsyn) [reply modules/compute/src/test/java/org/apache/ignite/internal/compute/loader/JobClassLoaderFactoryTest.java:69]: Is it better and why? I'm trying to keep the changes to a minimum here, with `join()` you can see easier that the test is the same.

File: modules/compute/src/main/java/org/apache/ignite/internal/compute/state/InMemoryComputeStateMachine.java [hunk 3]
```diff
@@ -139,6 +139,8 @@ private void changeJobState(UUID jobId, Function<JobState, JobState> newStateFun
             JobState currentState = currentStatus.state();
             JobState newState = newStateFunction.apply(currentState);
 
+            System.out.println("changeJobState currentState: " + currentState + ", newState: " + newState);
```
[the remark this answers is not in the document]
Comment (ptupitsyn) [reply modules/compute/src/main/java/org/apache/ignite/internal/compute/state/InMemoryComputeStateMachine.java]: Good catch, removed.

File: modules/compute/src/test/java/org/apache/ignite/internal/compute/loader/JobClassLoaderFactoryTest.java [hunk 4]
```diff
@@ -90,8 +91,8 @@ public void unit1BothVersions() throws Exception {
 
             // and classes are loaded in the aplhabetical order
             ComputeJob<Integer> job1 = (ComputeJob<Integer>) unitJobClass.getDeclaredConstructor().newInstance();
-            Integer result1 = job1.execute(null);
-            assertEquals(1, result1);
+            CompletableFuture<Integer> result1 = job1.executeAsync(null);
+            assertEquals(1, result1.join());
```
[the remark this answers is not in the document]
Comment (ptupitsyn) [reply modules/compute/src/test/java/org/apache/ignite/internal/compute/loader/JobClassLoaderFactoryTest.java]: Fixed.

File: modules/compute/src/test/java/org/apache/ignite/internal/compute/loader/JobClassLoaderFactoryTest.java [hunk 5]
```diff
@@ -64,14 +65,14 @@ public void unit1() throws Exception {
...
 
             // and classes from the second unit are loaded from the second class loader
             Class<?> clazz2 = classLoader2.loadClass(UNIT_JOB_CLASS_NAME);
             ComputeJob<String> job2 = (ComputeJob<String>) clazz2.getDeclaredConstructor().newInstance();
-            String result2 = job2.execute(null);
-            assertEquals("Hello World!", result2);
+            CompletableFuture<String> result2 = job2.executeAsync(null);
+            assertEquals("Hello World!", result2.join());
```
[the remark this answers is not in the document]
Comment (ptupitsyn) [reply modules/compute/src/test/java/org/apache/ignite/internal/compute/loader/JobClassLoaderFactoryTest.java]: Fixed.

File: modules/compute/src/test/java/org/apache/ignite/internal/compute/loader/JobClassLoaderFactoryTest.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (ptupitsyn) [reply modules/compute/src/test/java/org/apache/ignite/internal/compute/loader/JobClassLoaderFactoryTest.java]: Fixed.

File: modules/compute/src/main/java/org/apache/ignite/internal/compute/task/TaskExecutionInternal.java [hunk 6]
```diff
@@ -93,7 +93,8 @@ public TaskExecutionInternal(
         splitExecution = executorService.submit(
                 () -> {
                     MapReduceTask<R> task = instantiateTask(taskClass);
-                    return new SplitResult<>(task, task.split(context, args));
+
+                    return completedFuture(new SplitResult<>(task, task.split(context, args)));
```
[the remark this answers is not in the document]
Comment (ptupitsyn) [reply modules/compute/src/main/java/org/apache/ignite/internal/compute/task/TaskExecutionInternal.java:97]: Yes: https://issues.apache.org/jira/browse/IGNITE-22512

File: modules/api/src/main/java/org/apache/ignite/compute/ComputeJob.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (ptupitsyn) [reply modules/api/src/main/java/org/apache/ignite/compute/ComputeJob.java:35]: The user can return `completedFuture()` or `null`, it does not make any difference. Let's keep it this way.

File: modules/client-handler/src/main/java/org/apache/ignite/client/handler/requests/table/ClientStreamerWithReceiverBatchSendRequest.java [hunk 7]
```diff
@@ -102,7 +102,7 @@ public static CompletableFuture<Void> process(
 
     private static class ReceiverRunnerJob implements ComputeJob<List<Object>> {
         @Override
-        public @Nullable List<Object> execute(JobExecutionContext context, Object... args) {
+        public CompletableFuture<List<Object>> executeAsync(JobExecutionContext context, Object... args) {
```
[the remark this answers is not in the document]
Comment (ptupitsyn) [reply modules/client-handler/src/main/java/org/apache/ignite/client/handler/requests/table/ClientStreamerWithReceiverBatchSendRequest.java:105]: Fixed

File: modules/client-handler/src/main/java/org/apache/ignite/client/handler/requests/table/ClientStreamerWithReceiverBatchSendRequest.java [hunk 8]
```diff
@@ -115,7 +115,7 @@ private static class ReceiverRunnerJob implements ComputeJob<List<Object>> {
 
             CompletableFuture<List<Object>> receiveFut = receiver.receive(receiverInfo.items(), receiverContext, receiverInfo.args());
 
-            return receiveFut == null ? null : receiveFut.join();
+            return receiveFut == null ? nullCompletedFuture() : receiveFut;
```
[the remark this answers is not in the document]
Comment (ptupitsyn) [reply modules/client-handler/src/main/java/org/apache/ignite/client/handler/requests/table/ClientStreamerWithReceiverBatchSendRequest.java]: Fixed

File: modules/client/src/test/java/org/apache/ignite/client/fakes/FakeCompute.java [hunk 9]
```diff
@@ -104,11 +104,12 @@ public <R> JobExecution<R> executeAsyncWithFailover(
...
-            Class<ComputeJob<Object>> jobClass = ComputeUtils.jobClass(this.getClass().getClassLoader(), jobClassName);
-            ComputeJob<Object> job = ComputeUtils.instantiateJob(jobClass);
-            Object jobRes = job.execute(new JobExecutionContextImpl(ignite, new AtomicBoolean(), this.getClass().getClassLoader()), args);
+            Class<ComputeJob<R>> jobClass = ComputeUtils.jobClass(this.getClass().getClassLoader(), jobClassName);
+            ComputeJob<R> job = ComputeUtils.instantiateJob(jobClass);
+            CompletableFuture<R> jobFut = job.executeAsync(
+                    new JobExecutionContextImpl(ignite, new AtomicBoolean(), this.getClass().getClassLoader()), args);
 
-            return jobExecution(completedFuture((R) jobRes));
+            return jobExecution(jobFut);
```
[the remark this answers is not in the document]
Comment (ptupitsyn) [reply modules/client/src/test/java/org/apache/ignite/client/fakes/FakeCompute.java:112]: Fixed.

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/runner/app/client/ItThinClientComputeTest.java [hunk 10]
```diff
@@ -737,39 +745,47 @@ private List<ClusterNode> sortedNodes() {
...
 
     private static class ExceptionJob implements ComputeJob<String> {
         @Override
-        public String execute(JobExecutionContext context, Object... args) {
-            throw new ArithmeticException("math err");
+        public CompletableFuture<String> executeAsync(JobExecutionContext context, Object... args) {
+            boolean asyncJob = args.length > 0 && (Boolean) args[0];
+
+            if (asyncJob) {
+                return CompletableFuture.failedFuture(new ArithmeticException("math err"));
```
[the remark this answers is not in the document]
Comment (ptupitsyn) [reply modules/runner/src/integrationTest/java/org/apache/ignite/internal/runner/app/client/ItThinClientComputeTest.java]: Fixed.

**Pull request merged**